### PR TITLE
Fix and enable sam4e ethernet port

### PR DESCRIPTION
**Pull request by nandojve**
This PR add all missing pieces to fix and enable SAM4E Ethernet port.

The changes were tested on SAM4E and SAMV71 boards using samples/net/telnet. Tests are conducted using DHCPv4, collecting stats and ping 8.8.8.8

### CODE REVIEW

**Review by stephanosio — COMMENTED** [review]
We should add `supported: netif:eth` in `sam4e_xpro.yaml` too.

**Comment by stephanosio** [comment]
Just a reminder that, if #23045 gets merged before this PR, we need to add device tree stuff here as well.

I already have the commits for that here:
https://github.com/stephanosio/zephyr/commit/a14216636747bf28107f7a7183ba75fec014752e
https://github.com/stephanosio/zephyr/commit/da8a86872adb1404363644c1138c34385acaf02b

**Comment by nandojve** [comment]
> Just a reminder that, if #23045 gets merged before this PR, we need to add device tree stuff here as well.
> 
> I already have the commits for that here:
> [stephanosio@a142166](https://github.com/stephanosio/zephyr/commit/a14216636747bf28107f7a7183ba75fec014752e)
> [stephanosio@da8a868](https://github.com/stephanosio/zephyr/commit/da8a86872adb1404363644c1138c34385acaf02b)

**Comment by nandojve** [comment]
@mnkp

**Review by jukkar — APPROVED** [review]
LGTM

**Comment by stephanosio** [comment]
#23045 has been merged. This PR can be rebased now.

**Comment by nandojve** [comment]
Rebased.

**Comment by nandojve** [comment]
Rebase and fixed flash_sam 9d6530c

File: dts/bindings/ethernet/atmel,sam-gmac.yaml [hunk 1]
```diff
@@ -15,3 +16,32 @@ properties:
       type: int
       required: true
       description: Number of hardware TX and RX queues
+
+    local-mac-address:
+      type: uint8-array
+      description:
+        Specifies the MAC address that was assigned to the network device
```
Comment (stephanosio) [inline dts/bindings/ethernet/atmel,sam-gmac.yaml:23]: `local-mac-address` is already defined in `ethernet.yaml`, so no need to re-define it here; though the property description there could probably use some improvement.

File: dts/bindings/ethernet/atmel,sam-gmac.yaml [hunk 2]
```diff
@@ -15,3 +16,32 @@ properties:
...
+          Specifies maximum speed in Mbit/s supported by the device.
+        default: 100
+
+    phy-connection-type:
+        type: string
+        description:
+          Operation mode of the PHY interface
+        enum:
+          - rmii
+          - mii
```
Comment (stephanosio) [inline dts/bindings/ethernet/atmel,sam-gmac.yaml:46]: Shouldn't this be:
```
- "rmii"
- "mii"
```
(cc @galak @ulfalizer)

File: dts/bindings/ethernet/atmel,sam-gmac.yaml (hunk 1 above)
Comment (nandojve) [reply dts/bindings/ethernet/atmel,sam-gmac.yaml]: indeed, fixed!

File: dts/bindings/ethernet/atmel,sam-gmac.yaml (hunk 2 above)
Comment (nandojve) [reply dts/bindings/ethernet/atmel,sam-gmac.yaml]: fixed!

**Review by stephanosio — APPROVED** [review]

**Review by nvlsianpu — APPROVED** [review]
approve to the flash driver patch

**Comment by nandojve** [comment]
rebase - flash fix was already merged.

File: drivers/ethernet/Kconfig.sam_gmac [hunk 3]
```diff
@@ -6,7 +6,7 @@
 menuconfig ETH_SAM_GMAC
 	bool "Atmel SAM Ethernet driver"
 	depends on SOC_FAMILY_SAM
-	select NOCACHE_MEMORY
+	select NOCACHE_MEMORY if !SOC_SERIES_SAM4E
```
Comment (vlotech) [inline drivers/ethernet/Kconfig.sam_gmac:9]: instead of using SAM4E, I would advise to use:
select NOCACHE_MEMORY if ARCH_HAS_NOCACHE_MEMORY_SUPPORT

Comment (nandojve) [reply drivers/ethernet/Kconfig.sam_gmac]: Done!

**Comment by nandojve** [comment]
rebase and revert dt_node_int_prop_int to Kconfig 694cd62 until a better solution. In all cases, path isn't a valid solution because SoCs can have different address space for same peripheral.

File: drivers/ethernet/Kconfig.sam_gmac [hunk 4]
```diff
@@ -15,7 +15,12 @@ if ETH_SAM_GMAC
 config ETH_SAM_GMAC_QUEUES
 	int "Number of active hardware TX and RX queues"
 	default 1
-	range 1 $(dt_node_int_prop_int,/soc/ethernet@40050088,num-queues)
+	range 1 1 if SOC_SERIES_SAM4E
+	range 1 3 if !SOC_SERIES_SAM4E && \
+		     !SOC_ATMEL_SAME70_REVB && \
+		     !SOC_ATMEL_SAMV71_REVB
+	range 1 6 if SOC_ATMEL_SAME70_REVB || \
+		     SOC_ATMEL_SAMV71_REVB
```
Comment (stephanosio) [inline drivers/ethernet/Kconfig.sam_gmac:23]: TODO: Investigate how to deal with this (cc @galak)
https://github.com/zephyrproject-rtos/zephyr/pull/24057#discussion_r406572496

Comment (stephanosio) [reply drivers/ethernet/Kconfig.sam_gmac:23]: @nandojve See #24256

**Comment by nandojve** [comment]
rebase and added:

9ad5426 better approach to handle __nocache
62cca3b better way to handle in Kconfig

Thank you @stephanosio @galak @vlotech for all effort spend with the Kconfig issue.

**Comment by nandojve** [comment]
Hi everyone, I was wondering if there is anything missing here? This is an important feature that have been blocked since Feb and will be great have it on Zephyr 2.3.

**Pull request merged**